CONTRIBUTING.md

```diff
@@ -1,61 +1,520 @@
+        this.merger = new Merger(localStorageAdapter, mutationOutbox);
+    }
+
+    // Name the test in a way that you can tell the developer's intention
+    // why did you decide to write this test? What should the code _do_?
+    @Test
+    public void mergerAppliesUpdates() {
+        // Arrange: document preconditions for this test here
+        ...
+
+        // Act: what is the action of this test?
+        ...
+        // Assert: Given the arrangement, how does the code respond to the action?
+    }
+}
+```
+
+These templates encode some standards and best practices.
+[Checkstyle](https://github.com/checkstyle/checkstyle/blob/master/README.md) and
+[Android Lint](https://developer.android.com/studio/write/lint#overview)
+are the ultimate authorities on this, and will tell you if something is
+badly formatted, or exciting a known shortcoming in Java, Android, or
+their library ecosystem.
+
+Some suggestions include:
+
+* All files get a copyright header with the year of creation;
+* Use [`@NonNull`](https://developer.android.com/reference/kotlin/androidx/annotation/NonNull)
+  and [`@Nullable`](https://developer.android.com/reference/kotlin/androidx/annotation/Nullable)
+  annotations on any/all params, so that
+  [Kotlin knows what to do](https://kotlinlang.org/docs/reference/java-interop.html#nullability-annotations),
+  when it consumes the library;
+* Label Arrange/Act/Assert portions of your unit tests, so others can follow
+  their flow;
+* Arrange shared dependencies and test objects in an
+  [`@Before`](http://junit.sourceforge.net/javadoc/org/junit/Before.html) method;
+* Prefer `private` and `final` class members where-ever possible. This limits scope
+  and mutability;
+* Add documentation to anything that says `public` or `protected`. The
+  checkstyle doesn't require documentation on package-local or `private` fields, but
+  that might be a good idea, too, if you're doing some interesting/non-trivial
+  work in those places.
 
-When filing an issue, please check [existing open](https://github.com/awslabs/amplify-android/issues), or [recently closed](https://github.com/awslabs/amplify-android/issues?utf8=%E2%9C%93&q=is%3Aissue%20is%3Aclosed%20), issues to make sure somebody else hasn't already
-reported the issue. Please try to include as much information as you can. Details like these are incredibly useful:
+### Build and Validate Your Work
 
-* A reproducible test case or series of steps
-* The version of our code being used
-* Any modifications you've made relevant to the bug
-* Anything unusual about your environment or deployment
+This will perform a clean build, run Checkstyle, Android Lint, and all unit
+tests. This must complete successfully before proposing a PR.
 
+```
+./gradlew clean build
+```
+
+### Run Instrumentation Tests
+
+The instrumentation tests presume the presence of various backend resources.
+Currently, the tests use "special" Amazon-only `amplifyconfiguration.json` files in
```

Comment:
The word special is doing a lot of work in this section. External contributors aren't expected to run these, so perhaps we should phrase it that way instead.

Reply:
Hm, true. I've removed this word, and replaced it with a more well-rounded depiction of what is special about the files.